src/main/java/mods/railcraft/common/carts/EntityCartHopper.java

```diff
@@ -140,38 +124,75 @@ public double getZPos() {
     public void onUpdate() {
         super.onUpdate();
 
-        if (Game.isHost(world) && this.isEntityAlive() && this.getBlocked()) {
-            BlockPos blockpos = new BlockPos(this);
+        if (Game.isHost(world) && !isDead) {
+            BlockPos nowPos = new BlockPos(this);
+            if (enabled) {
+                // cooldown when transferring from a same block
+                if (Objects.equals(lastPosition, nowPos)) {
+                    if (transferCooldown > 0)
+                        transferCooldown--;
+                } else {
+                    transferCooldown = 0;
+                }
 
-            if (blockpos.equals(this.lastPosition)) {
-                --this.transferTicker;
-            } else {
-                this.setTransferTicker(0);
+                if (transferCooldown <= 0) {
+                    if (transferAndNeedsCooldown()) {
+                        transferCooldown = 4;
+                    }
+                }
             }
 
-            if (!this.canTransfer()) {
-                this.setTransferTicker(0);
+            if (pushCounter == 0) {
+                tryPushItem();
+            }
 
-                if (this.captureDroppedItems()) {
-                    this.setTransferTicker(4);
-                    this.markDirty();
+            pushCounter++;
+            if (pushCounter >= 5)
+                pushCounter = 0;
+            lastPosition = nowPos;
+        }
+    }
+
+    private void tryPushItem() {
+        boolean emptySlot = false;
+        // Push full stacks whenever possible
+        for (IExtInvSlot slot : InventoryIterator.get(this)) {
+            if (slot.getStack().getCount() == slot.maxStackSize()) {
```

Comment:
You might want to consider the ItemStack max stack size well, not just the slot max capacity.